src/ray/raylet/local_task_manager.cc

```diff
@@ -138,6 +138,87 @@ void LocalTaskManager::DispatchScheduledTasksToWorkers() {
     }
     auto &sched_cls_info = info_by_sched_cls_.at(scheduling_class);
 
+    // Fair scheduling is applied only when the total CPU requests exceed the node's
+    // capacity. This skips scheduling classes whose number of running tasks exceeds the
+    // average number of tasks per scheduling class.
+
+    // The purpose of fair scheduling is to ensure that each scheduling class has an
+    // equal chance of being selected for dispatch. For instance, in a pipeline with both
+    // data producers and consumers, we aim for consumers to have the same chance to be
+    // dispatched as producers. This prevents memory peak caused by dispatching all
+    // producer tasks first. Example: Instead of dispatching [f,f,f], we want to dispatch
+    // [f,f,g], where 'f' and 'g' represent different scheduling classes. Here, we
+    // prioritize dispatching 'g' to maintain fairness.
```

Comment:
This example made me confused on why the tasks to dispatch are different in the comparison. What about:

    Example: when we have 3 CPUs and 2 scheduling classes `f` and `g` each with 4 tasks, we dispatch [f,f,g] instead of [f,f,f].

Reply:
We may also need a explicit, verifiable rule, like:

    A scheduling class is skipped from scheduling, if its fair_share exceeds the average fair_share among all scheduling classes. fair_share is number of running tasks with >0 CPU.

Reply:
Thanks! I just make it more clear based on you comments!